README.ideal.md

```diff
@@ -0,0 +1,131 @@
+
+1. A unified configuration framework for all of these
+2. A powerful and ergonomic language to allow you to easily abstract over these configurations, with discoverability and early error reporting
+3. A way to get these tools to communicate with each others, so that your custom scripts can easily leverage your package manager to get the dependencies they need, your CI setup can understand your build system and parallelise efficiently, and so on.
+
+## Getting started
+
+To start using `nickel-nix`, you need `Nix` to be installed and configured.
+If it isn't already the case, you can get it with:
+
+```console
+$ curl -L https://nixos.org/nix/install | bash
+# We also need a couple of experimental Nix features
+$ mkdir -p ~/.config/nix
+$ echo 'experimental-features = nix-command flakes' >> ~/.config/nix/nix.conf
+```
+
+Then bootstrap a project with:
+
+```console
+$ nix flake init -t github:nickel-lang/nickel-nix
+# Edit the project file to fit your needs
+$ $EDITOR project.ncl
+# Enter the environment
+$ . .nickel-nix/activate
+```
+
+## Managing your environment with Nixel
+
+### Dependencies
+
+Nixel can be used to declare the dependencies for your project.
+These can then be instantiated using [Nix](https://nixos.org/nix).
+
+> TODO: Document
+
+### CI configuration
+
+You can use Nixel to generate a CI configuration.
+This is extremely handy if you want to test some non-trivial matrix of platforms/configuration.
+
+Only Github actions is supported at the moment, but we plan to add more options soon.
+
+> TODO: Document
+
+### Development services
+
+Nixel can also be used for some simple service management.
+In combination with the Nix integration, this allows you to quickly spawn any service that you might depend on.
+
+Once the services have been defined in your `project.ncl` file, you can run `nixel up` to start them.
+
+> TODO: Document
+
+### Deployment
+
+If you want to push things yet a bit further, Nixel can also interoperate with [tf-ncl](https://github.com/tweag/tf-ncl) to allow you to manage the deployment infrastructure for your project.
+
+> TODO: Document
+
+## Structure of the `project.ncl` file
+
+`project.ncl` is the main entrypoint for `nickel-nix`.
+It describes the packages that should be included in your development environment, as well as the template for any file you'd like to generate:
+
+```nickel
+let Nixel = import ".nickel-nix/lock.ncl" in
+{
+  nix_inputs.spec = {
+    pandoc = {},
+    make = {},
+    prettier = {},
+    redis = {},
+  },
+  shells.build =
+    {
+      packages = [nix_inputs.pandoc],
+      scripts = {
+        build = nix-s%"
+                #!/usr/bin/env bash
+
+                %{nix_inputs.make}/bin/make -j$(nproc) -l$(nproc)
+            "%,
+      },
+    }
+    & Nixel.shells.Rust
+    & Nixel.shells.Nickel,
+  shells.dev =
+    shells.build
+    & {
+      env = {
+        DEV_ENDPOINT = "http://localhost:1234",
+      },
+      packages = [nix_inputs.prettier],
+    },
```

Comment:
So are we abandoning the idea that, say, `Nixel.shells.Rust` provides a `build` and a `dev` shell? We probably shouldn't. Here is how I'd rewrite this

```suggestion
  shells = Nixel.shells.Rust,
  shells = Nixel.shells.Nickel,
  
  shells.build =
    {
      packages = [nix_inputs.pandoc],
      scripts = {
        build = nix-s%"
                #!/usr/bin/env bash

                %{nix_inputs.make}/bin/make -j$(nproc) -l$(nproc)
            "%,
      },
    },
    
  shells.dev =
    {
      env = {
        DEV_ENDPOINT = "http://localhost:1234",
      },
      packages = [nix_inputs.prettier],
    },
```

There's an implicit assumption that `shells.dev` always contains `shells.build` in this, whatever we add there. It may be subtle to attain, but it's definitely worth (maybe `shells.dev` doesn't actually contain `shells.build`, maybe there's a field containing a list/record of super-shells, and when you call `nix develop .#dev` it does merge the shell with its super-shells). Remember to always keep the boilerplate to a minimum. The easiest thing to type should be the right default.

Reply:
> So are we abandoning the idea that, say, Nixel.shells.Rust provides a build and a dev shell?

Not in the intent, no.
I'm still not clear on the exact shape we want for these definitions. But yeah, your version seems better